src/range-slider/index.tsx

```diff
@@ -63,281 +62,253 @@ export interface RangeSliderProperties extends ThemedProperties {
-			step = 1,
-			widgetId = this._widgetId,
-			onFocus,
-			onBlur
-		} = this.properties;
-		const { minName = `${name}_min`, maxName = `${name}_max` } = this.properties;
+export interface RangeSliderICache {
+	initialValue?: RangeValue;
+	value?: RangeValue;
+}
 
-		return {
-			...formatAriaProperties(aria),
-			'aria-invalid': valid === false ? 'true' : null,
-			'aria-readonly': readOnly === true ? 'true' : null,
-			'aria-describedby': isSlider1 ? this._minLabelId : this._maxLabelId,
-			'aria-labelledby': `${widgetId}-label`,
-			type: 'range',
-			min: `${min}`,
-			max: `${max}`,
-			step: `${step}`,
-			readonly: readOnly,
-			required,
-			disabled,
-			name: isSlider1 ? minName : maxName,
-			onblur: () => {
-				onBlur && onBlur();
-			},
-			onfocus: () => {
-				onFocus && onFocus();
-			},
-			oninput: (event: Event) => {
-				this._onInput(event, isSlider1);
-			},
-			classes: [this.theme(css.input), fixedCss.nativeInput]
-		};
-	}
+const factory = create({
+	dimensions,
+	focus,
+	icache: createICacheMiddleware<RangeSliderICache>(),
+	theme
+}).properties<RangeSliderProperties>();
 
-	protected renderOutput(value: RangeValue, percentValue: number[]): DNode {
-		const { output, outputIsTooltip = false } = this.properties;
+export const RangeSlider = factory(function RangeSlider({
+	id,
+	middleware: { dimensions, focus, icache, theme },
+	properties
+}) {
+	const { name = '', max: maxRestraint = 100, min: minRestraint = 0 } = properties();
 
-		const outputNode = output ? output(value) : `${value.min}, ${value.max}`;
+	const {
+		aria = {},
+		classes,
+		disabled,
+		label,
+		labelAfter,
+		labelHidden,
+		maxName = `${name}_max`,
+		maximumLabel = 'Maximum',
+		minName = `${name}_min`,
+		minimumLabel = 'Minimum',
+		onBlur,
+		onFocus,
+		onOut,
+		onOver,
+		onValue,
+		output,
+		outputIsTooltip,
+		readOnly,
+		required,
+		showOutput = false,
+		step = 1,
+		theme: themeProp,
+		valid,
+		initialValue = {
+			max: maxRestraint,
+			min: minRestraint
+		},
+		widgetId = `range-slider-${id}`
+	} = properties();
 
-		// output styles
-		let outputStyles: { left?: string; top?: string } = {};
-		if (outputIsTooltip) {
-			outputStyles = {
-				left: `${Math.round(
-					(percentValue[0] + (percentValue[1] - percentValue[0]) / 2) * 100
-				)}%`
-			};
-		}
+	let value = icache.get('value');
+	const existingInitialValue = icache.get('initialValue');
 
-		return v(
-			'output',
-			{
-				classes: this.theme([css.output, outputIsTooltip ? css.outputTooltip : null]),
-				for: this._widgetId,
-				styles: outputStyles,
-				tabIndex: -1 /* needed so Edge doesn't select the element while tabbing through */
-			},
-			[outputNode]
-		);
+	if (
+		!existingInitialValue ||
+		initialValue.min !== existingInitialValue.min ||
+		initialValue.max !== existingInitialValue.max
+	) {
+		icache.set('value', initialValue);
+		icache.set('initialValue', initialValue);
+		value = initialValue;
+		onValue && onValue(initialValue);
```

Comment:
why are you calling `onValue` here? I've noticed we're doing this elsewhere too but i'm not sure why.

Reply:
If a consumer sets a new `initialValue` on this widget, this condition will be triggered. I think `onValue` is called here since the new `initialValue` will in turn result in a new internal `value` being set. Even though it's not technically needed, it would feel odd if this didn't trigger `onValue`, but all other internal value changes did.

Reply:
i'm not sure we need to call `onValue` as the value was just set, could easily cause a loop

Reply:
Updated.